west/runner/core.py

```diff
@@ -256,15 +256,21 @@ class ZephyrBinaryRunner(abc.ABC):
     - 'flash': flash a previously configured binary to the board,
       start execution on the target, then return.
 
-    - 'debug': connect to the board via a debugging protocol, then
-      drop the user into a debugger interface with symbol tables
-      loaded from the current binary, and block until it exits.
+    - 'debug': connect to the board via a debugging protocol, program
+      the flash, then drop the user into a debugger interface with
+      symbol tables loaded from the current binary, and block until it
+      exits.
 
     - 'debugserver': connect via a board-specific debugging protocol,
```

Comment:
I've always been confused about the way different runners do or don't program the flash in the 'debug' command, so I think making the distinction explicit is worth doing. I'm not sure how it affects the 'debugserver' command though -- is it always meant to program the flash, like 'debug'? Or is it like 'attach'?

Reply:
Related: is there a need for an "attachserver"?

Reply:
A number of the runners (arc, jlink, nsim, pyocd) program the flash in the gdb client using the argument `-ex load`. There a few (intel_s1000, nios2, openocd, xtensa) that I'm not sure if/how they program the flash. The rest (bossac, dfu, esp32, nrfjprog) only support the `flash` command, no debugging.

If we agree that `debugserver` does not program the flash, then there is no need for an `attachserver`

Reply:
This is what I think you're saying the desired behavior is:

| Command     | Reset target | Reflash target | Attach debugger | Run target code |
|-------------|--------------|----------------|-----------------|-----------------|
| debugserver | Y            | N              | N               | N               |
| attach      | Y            | N              | Y               | N               |
| debug       | Y            | Y              | Y               | N               |
| flash       | Y            | Y              | N               | Y               |

That seems reasonable and I agree there's no need for attachserver this way.

It would be nice to add something like that table to the West documentation when this code moves into Zephyr, now that the debug* vs. flash nomenclature is being mixed up by "attach".

(If I were doing this from scratch I would have "debug" behave like "attach", but we seem to have inherited this behavior from the original RIOT shell scripts.)